### ISSUE TITLE
Bug with c++ syntax highlighting on Gitlab

### ISSUE THREAD

**Issue by n-ham** [rouge-ruby/rouge#1009]
### Summary

The C++ syntax highlighter for Gitlab.com (which is apparently Rogue) is getting confused about what is and is not a string, it is also highlighting 'if' as red sometimes when it should not.

Also, it would be nice to add highlighting for stuff like std::istream, std::ifstream, std::ostream, std::ofstream, std::pair, std::set, std::string, std::vector, std::queue, std::priority_queue, std::deque, std::endl, std::map, std::greater, std::less, etc.. 

### Steps to reproduce

1. Create a new repository on Gitlab
2. Push a .cc file containing the offending code (available from the example project linked below)
3. View bug on the repository webpage

### Example Project

An example project has been set up [here](https://gitlab.com/n-ham-paper-files/cpp-syntax-highlighting-bug). See the `README.md` file for an explanation of the bugs throughout the included files. 

### What is the current *bug* behavior?

The bug first appeared when I uploaded `algorithms.h` to Gitlab. From what I can tell the problems begin on line 468 with a red 'if', and then on line 469 it gets confused about what is and is not a string midway through ") {$A_", even though ") {$A_" is handled correctly on line 467.

### What is the expected *correct* behavior?

Check from line 468 onward in `algorithms-d.h` to see an example of syntax highlighting with the same code that does not have any bugs.

The files `algorithms-c.h`, `algorithms-cpp.h`, `algorithms-d.h` and `algorithms-java.h` have syntax highlighting for c, cpp, d and java respectively as specified in the `.gitattributes` file. The bug is present for c and cpp, and not present for d and java. 

There is more information about the errors in the README.md file for the [example project](https://gitlab.com/n-ham-paper-files/cpp-syntax-highlighting-bug).

Also, it would be nice to add highlighting for stuff like std::istream, std::ifstream, std::ostream, std::ofstream, std::pair, std::set, std::string, std::vector, std::queue, std::priority_queue, std::deque, std::endl, std::map, std::greater, std::less, etc.. 

### Output of checks

This bug happens on Gitlab.com

**Comment by dblessing** [comment]
@n-ham Thanks for reporting this. I did some debugging and it seems the `else if` is getting caught by the 'functions' rule in `c.rb` in the `:root` state. If I move the statement rule up above this rule, just below `mixin :expr_whitespace` things seem to work for me. I'm not fond of this as that rule is going to catch everything, but if it works, then I guess...

Can you please test this change against some of your files to see if it breaks anything? Let me know if you have trouble figuring out how to test this on GitLab.

**Comment by dblessing** [comment]
For completeness:

## Current

```ruby
      state :root do
        mixin :expr_whitespace

        # functions
        rule %r(
          ([\w*\s]+?[\s*]) # return arguments
          (#{id})          # function name
          (\s*\([^;]*?\))  # signature
          (#{ws})({)         # open brace
        )mx do |m|
          # TODO: do this better.
          recurse m[1]
          token Name::Function, m[2]
          recurse m[3]
          recurse m[4]
          token Punctuation, m[5]
          push :function
        end

        # function declarations
        rule %r(
          ([\w*\s]+?[\s*]) # return arguments
          (#{id})          # function name
          (\s*\([^;]*?\))  # signature
          (#{ws})(;)       # semicolon
        )mx do |m|
          # TODO: do this better.
          recurse m[1]
          token Name::Function, m[2]
          recurse m[3]
          recurse m[4]
          token Punctuation, m[5]
          push :statement
        end
        
        rule(//) { push :statement }
      end
```

## Proposed

```ruby      
     state :root do
        mixin :expr_whitespace

        rule(//) { push :statement }

        # functions
        rule %r(
          ([\w*\s]+?[\s*]) # return arguments
          (#{id})          # function name
          (\s*\([^;]*?\))  # signature
          (#{ws})({)         # open brace
        )mx do |m|
          # TODO: do this better.
          recurse m[1]
          token Name::Function, m[2]
          recurse m[3]
          recurse m[4]
          token Punctuation, m[5]
          push :function
        end

        # function declarations
        rule %r(
          ([\w*\s]+?[\s*]) # return arguments
          (#{id})          # function name
          (\s*\([^;]*?\))  # signature
          (#{ws})(;)       # semicolon
        )mx do |m|
          # TODO: do this better.
          recurse m[1]
          token Name::Function, m[2]
          recurse m[3]
          recurse m[4]
          token Punctuation, m[5]
          push :statement
        end

      end
```

**Comment by vidarh** [comment]
Looks like there are actually multiple different problems here.

1. ws is not made optional before ; and { respectively:

```c
   void foo();
   void foo(){}
```
2. Pushing :statement in the function declaration is broken. It means subsequent function declarations or definitions does not work properly even if they have the needed whitespace:

```c
  void foo() ;
  void bar() ;
```
(note the whitespace before `;` is significant because of no 1)

With the above fixed, *this* works:

```c
  void foo() {
    if(x) {
    } else if(y) {
    }
```

3. The function name/declaration thing causes problems mainly because :root the pushes :statement, so that if anything is wrong with a function declaration, you don't get another chance until you've lucked out and triggered enough :pop!'s from :statement. This is brittle. 

I think rewriting :root to this seems to make it a lot more robust (I also combined function declarations and definitions, making it quite a bit shorter)

```ruby
      state :root do
        mixin :expr_whitespace
        rule %r(
          ([\w*\s]+?[\s*]) # return arguments
          (#{id})          # function name
          (\s*\([^;]*?\))  # signature
          (#{ws}?)({|;)    # open brace or semicolon
        )mx do |m|
          # TODO: do this better.                                                 
          recurse m[1]
          token Name::Function, m[2]
          recurse m[3]
          recurse m[4]
          token Punctuation, m[5]
          if m[5] == ?{
            push :function
          end
        end
        mixin :statements
        rule /\{/, Punctuation, :function
      end
```
+ adding ";" to the punctuation set in :statements

And we can then delete :statement entirely. I'll do some more testing before I prepare a pull-request. If you spot any issues let me know.

**Comment by vidarh** [comment]
The above will also fix #879 I believe.

**Comment by JBetser** [comment]
There are two similar bugs with macros:
#define PPCONCAT(A, B) \
    A##B
'\' and '##' get incorrectly highlighted in Gitlab

**Comment by pyrmont** [comment]
Fixed by #1069. Thanks @vidarh!
